backend/config.go

```diff
@@ -151,6 +156,76 @@ func (c *GrafanaCfg) ConcurrentQueryCount() (int, error) {
 	return i, nil
 }
 
+type SqlDefaults struct {
+	MaxOpenConns           int
+	MaxIdleConns           int
+	MaxConnLifetimeSeconds int
+}
+
+func (c *GrafanaCfg) SqlDefaults() (SqlDefaults, error) {
```

Comment:
IMO it would be nicer to have as `config.SQL().<>` for the SQL configs as right now it's a bit confusing. I'd also consider returning `(int/string/etc., bool)` instead of `(int/string/etc., error)` for the field value functions (MaxOpenConns, MaxIdleConns, MaxConnLifetimeSeconds, RowLimit) unless you really want people to have to handle the errors / potentially error out. Same goes for `LogUserFacingDefaultError` in that case.

Reply:
@wbrowne 
thanks for the feedback... just to clarify some points there...:
> IMO it would be nicer to have as config.SQL().<> for the SQL configs as right now it's a bit confusing.

so you'd recommend to call them like this?
```
config.SQL().RowLimit()
config.SQL().MaxOpenConnsDefault()
config.SQL().MaxIdleConnsDefault()
config.SQL().MaxConnLifetimeSecondsDefault()
```
?

(the word `default` is important here. these are not the real limits. these are limits that will apply IF the datasource-instance does not have them set. wein mysql we set these values in a structure, and then we unmarshal the json into the same structure: https://github.com/grafana/grafana/blob/ff916d9c15d38eead1ed93c4da2d7fc0c5e3b081/pkg/tsdb/mysql/mysql.go#L56-L67

> consider returning (int/string/etc., bool) instead of (int/string/etc., error)

i see your point, but... if i understand this correctly, we do this so that people can ignore handling the `bool`, right? that should mean that the default-value of the function is useful... right now that is not the case. for example, the various sql-numbers, if they will be zero, i'm not sure what they will cause....still, i agree that it would be great if the plugins could just ask for the config-value and get something usable. i was also thinking about setting default-values in plugin-sdk-go, but then we are duplicating the code from `conf/defaults.ini` in grafana. WDYT?

Reply:
@wbrowne 
> > the word default is important here. these are not the real limits. these are limits that will apply IF the datasource-instance does not have them set.
> 
> So how does it work - these values are only used when they're not configured in the jsonData?

yes.
i think what happened is that this was a hardcoded default, and then last year sometime we changed the value, and added the possibility to change it, probably for backward-compatibility?  note 100% sure ( https://github.com/grafana/grafana/pull/65834 )

> 
> > we do this so that people can ignore handling the bool, right?
> 
> It's more about when an error is returned, people's natural response is to do:
> 
> ```go
> if err != nil {
>    return err
> }
> ```
> 
> whereas a `bool` forces them to think a bit more about it (at least I hope 😅 )

oh, my mistake. i somehow assumed people can just call it with receiving a single value, like how some builtin go structures do it, but here they would be forced to receive the boolean 👍 .

though... sorry to complicate this :) .. so for both "value missing" and "value cannot be converted to a number" we go with `ok=false`?

> 
> > i was also thinking about setting default-values in plugin-sdk-go, but then we are duplicating the code from conf/defaults.ini in grafana. WDYT?
> 
> Not too sure about this one. Maybe since these are plugin specific config, the plugin SDK can be now the source of truth for the default values instead of Grafana itself? Would like to get other people's thoughts on this.

good idea.

> Then I guess what gets passed in the request will be the actual config value and not defaults? 😕

i don't know if we can simplify it that much (would love to of course 😁 ).. this would require the datasource-plugin to store these values in it's `jsonData` at a "standard" location, right?

this whole thing is a bit frustrating, becuase no non-core datasource uses these settings (for obvious reasons), so it's hard to differentiate between "what is right" and "what is the current way". right now only mysql/mssql/postgres use these values, and they all store it at the same place in the json-data, so for these 3 it would work in an automatic way.

Reply:
thanks for the feedback @andresmgot @wbrowne . i think we discussed sql-defaults throughly, but i wonder how we feel about the two remaining ones:
- rowLimit
- userFacingDefaultError

should we do the same as `sqlDefaults()` , as in return a `(int, error)` for `rowLimit` and `(string, error)` for `userFacingDefaultError`?

Reply:
`userFacingDefaultError` should be its own separate method. If `rowLimit` doesn't belong to the SQL defaults group, then I'd be inclined to go back to my original suggestion IE

```go
config.SQL().RowLimit() 
config.SQL().MaxOpenConnsDefault()
```

Reply:
yes, I though `rowLimit` is also a SQL default config so it would go in the same struct

Reply:
i adjusted the code, right now it's like this:
```go
errorMsg, err := config.LogUserFacingDefaultError()
if err != nil {
  return err
}

sqlConfig, err := config.SQL()
if err != nil {
  return err
}
// sqlConfig.RowLimit int64 ( it's int64 in the grafana config)
// sqlConfig.DefaultMaxOpenConns int
// sqlConfig.DefaultMaxConnLifetimeSeconds int
// sqlConfig.DefaultMaxIdleConns int

// i moved the word `default` to the beginning of the variable name, i think it looks better that way
```

WDYT? @wbrowne  @andresmgot

Reply:
LGTM! One thing is a bit confusing -> `LogUserFacingDefaultError`. Should it just be `UserFacingDefaultError`?